Add experimental features flag option for tests

neetikasinghal opened the pull request:
### Description
As of now, there is no way in the tests to turn on the experimental feature flags for the tests, this pr aims to address it.

### Issues Resolved
Related prs: https://github.com/opensearch-project/opensearch-api-specification/pull/368

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock commented on .github/workflows/test-spec.yml:
```diff
@@ -26,7 +26,10 @@ jobs:
+          - version: 2.16.0
+            hub: opensearchstaging
+            ref: '@sha256:50fbfe3b95c41e92a113ada3e80513ba4524dfc8a25dc6aaeff2bbe1e1145d5f'
+            features: '-Dopensearch.experimental.feature.tiered_remote_index.enabled=true'
```
This feels like just a parameter, either make this `opts` (for `OPENSEARCH_JAVA_OPTS`) or (this might be harder but nicer)

```yaml
experimental_features:
  - tiered_remote_index: true
```

neetikasinghal replied:
updated to opts

dblock commented on .github/opensearch-cluster/docker-compose.yml:
```diff
@@ -8,6 +8,7 @@ services:
       - '9600:9600'
     environment:
       - 'OPENSEARCH_INITIAL_ADMIN_PASSWORD=${OPENSEARCH_PASSWORD:-myStrongPassword123!}'
+      - 'OPENSEARCH_JAVA_OPTS=${OPENSEARCH_EXPERIMENTAL_FEATURES}'
```
The list of opts is not a list of experimental features, so this should be `OPENSEARCH_JAVA_OPTS=${OPENSEARCH_JAVA_OPTS}`.

neetikasinghal replied:
updated

dblock commented on .github/workflows/test-spec.yml:
```diff
@@ -36,6 +39,7 @@ jobs:
       OPENSEARCH_DOCKER_REF: ${{ matrix.entry.ref }}
       OPENSEARCH_VERSION: ${{ matrix.entry.version }}
       OPENSEARCH_PASSWORD: ${{ matrix.entry.admin_password || 'myStrongPassword123!' }}
+      OPENSEARCH_EXPERIMENTAL_FEATURES: ${{ matrix.entry.opts || 'null' }}
```
This should be `OPENSEARCH_JAVA_OPTS` because it's not just a list of experimental features. 

Does it default to a string "null"? That's not the same thing as undefined. IMO it should read this:

```
OPENSEARCH_JAVA_OPTS: ${{ matrix.entry.opts }}
```

neetikasinghal replied:
will update this

dblock commented on CHANGELOG.md:
```diff
@@ -55,6 +55,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Added `plugins` to NodeInfoSettings ([#442](https://github.com/opensearch-project/opensearch-api-specification/pull/442)) 
 - Added test coverage ([#443](https://github.com/opensearch-project/opensearch-api-specification/pull/443))
 - Added `--opensearch-version` to `merger` that excludes schema elements per semver ([#428](https://github.com/opensearch-project/opensearch-api-specification/pull/428))
+- Added experimental features flag option for tests ([#454](https://github.com/opensearch-project/opensearch-api-specification/pull/454))
```
"Added passing OPENSEARCH_JAVA_OPTS into the docker container used for tests"

dblock reviewed (commented):
The code passes generic `OPENSEARCH_JAVA_OPTS` but claims this is `OPENSEARCH_EXPERIMENTAL_FEATURES` which it's not, it can be any opts. Get rid of mentions of "experimental features".

An alternate and likely a nicer implementation could be to actually pass a list of experimental features.

In test-spec.yml.

```yaml
- version: 2.16.0
   features:
     - tiered_remote_index: true
```

To convert this into `OPENSEARCH_JAVA_OPTS` you could use [join](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/expressions#join).

neetikasinghal commented:
> The code passes generic `OPENSEARCH_JAVA_OPTS` but claims this is `OPENSEARCH_EXPERIMENTAL_FEATURES` which it's not, it can be any opts. Get rid of mentions of "experimental features".
> 
> An alternate and likely a nicer implementation could be to actually pass a list of experimental features.
> 
> In test-spec.yml.
> 
> ```yaml
> - version: 2.16.0
>    features:
>      - tiered_remote_index: true
> ```
> 
> To convert this into `OPENSEARCH_JAVA_OPTS` you could use [join](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/expressions#join).

I tried the following:

```
    steps:
      - name: Set OPENSEARCH_JAVA_OPTS
        run: |
          # Initialize options array
          opts=()
          prefix="-Dopensearch.experimental.feature"
          suffix="enabled"
          
          # Extract features directly from the matrix
          features=${{ matrix.entry.features }}
          
          # Split features into key-value pairs
          IFS=',' read -r -a feature_pairs <<< "$features"
          
          # Process each key-value pair
          for pair in "${feature_pairs[@]}"; do
            IFS='=' read -r key value <<< "$pair"
            opts+=("${prefix}.${key}.${suffix}=${value}")
          done
          
          # Set the environment variable
          echo "OPENSEARCH_JAVA_OPTS=$(IFS=','; echo "${opts[*]}")" >> $GITHUB_ENV
          echo $(cat $GITHUB_ENV)
```

However, its coming as `OPENSEARCH_JAVA_OPTS=-Dopensearch.experimental.feature.Array.enabled=`

refer: https://github.com/opensearch-project/opensearch-api-specification/actions/runs/10189736051/job/28188313025

So, I am going to go with opts for now unless u have another suggestion.

dblock commented:
Looks like the update SHA is bringing in a new feature that's not documented in the spec. Either add it (looks easy), or change the SHA back to a previous one for now.

dblock commented:
I updated the 2.16 ref to the latest in #459 and tests didn't fail the way they did here, so I looked at the `@sha256:50fbfe3b95c41e92a113ada3e80513ba4524dfc8a25dc6aaeff2bbe1e1145d5f` SHA and locally all my tests passed. So I added `-Dopensearch.experimental.feature.tiered_remote_index.enabled=true` and it failed like in this PR. Looks like the `cache_utilized` and `file_cache` properties are only enabled with this experimental feature enabled.

dblock commented on .github/opensearch-cluster/docker-compose.yml:
```diff
@@ -8,6 +8,7 @@ services:
       - '9600:9600'
     environment:
       - 'OPENSEARCH_INITIAL_ADMIN_PASSWORD=${OPENSEARCH_PASSWORD:-myStrongPassword123!}'
+      - 'OPENSEARCH_JAVA_OPTS=${OPENSEARCH_JAVA_OPTS}'
```
In my experiments these quotes are unnecessary on both `OPENSEARCH_JAVA_OPTS` and `OPENSEARCH_INITIAL_ADMIN_PASSWORD`, so you can remove them.

neetikasinghal commented:
> I updated the 2.16 ref to the latest in #459 and tests didn't fail the way they did here, so I looked at the `@sha256:50fbfe3b95c41e92a113ada3e80513ba4524dfc8a25dc6aaeff2bbe1e1145d5f` SHA and locally all my tests passed. So I added `-Dopensearch.experimental.feature.tiered_remote_index.enabled=true` and it failed like in this PR. Looks like the `cache_utilized` and `file_cache` properties are only enabled with this experimental feature enabled.

Looks like its coming from here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/node/Node.java#L385 - this doesnt look right.

dblock commented:
> > I updated the 2.16 ref to the latest in #459 and tests didn't fail the way they did here, so I looked at the `@sha256:50fbfe3b95c41e92a113ada3e80513ba4524dfc8a25dc6aaeff2bbe1e1145d5f` SHA and locally all my tests passed. So I added `-Dopensearch.experimental.feature.tiered_remote_index.enabled=true` and it failed like in this PR. Looks like the `cache_utilized` and `file_cache` properties are only enabled with this experimental feature enabled.
> 
> Looks like its coming from here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/node/Node.java#L385

Are you looking into it? Open an issue at least?

dblock commented:
> > > > I updated the 2.16 ref to the latest in #459 and tests didn't fail the way they did here, so I looked at the `@sha256:50fbfe3b95c41e92a113ada3e80513ba4524dfc8a25dc6aaeff2bbe1e1145d5f` SHA and locally all my tests passed. So I added `-Dopensearch.experimental.feature.tiered_remote_index.enabled=true` and it failed like in this PR. Looks like the `cache_utilized` and `file_cache` properties are only enabled with this experimental feature enabled.
> > > 
> > > 
> > > Looks like its coming from here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/node/Node.java#L385
> > 
> > 
> > Are you looking into it? Open an issue at least?
> 
> yeah checking with @andrross to understand his thought process behind this change - doesnt look super right to me as If the tiered remote index feature flag is on and there is not search node it the cluster it would set the NODE_SEARCH_CACHE_SIZE_SETTING to 80%.

For this PR, remove `opts: -Dopensearch.experimental.feature.tiered_remote_index.enabled=true` from CI and it will go green, so we can then merge it.

In your other PR where you actually need `-Dopensearch.experimental.feature.tiered_remote_index.enabled=true` we need a fix. If it's in OpenSearch we can wait for it.

dblock commented:
The intermittent failure is an OOM in the sql plugin, I changed the threshold in another PR.
